org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/resource/BatchLinkableResourceStorageLoadable.java

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (c) 2014, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.xbase.resource;
+
+import java.io.BufferedInputStream;
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.zip.ZipInputStream;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.InternalEObject;
+import org.eclipse.emf.ecore.resource.impl.BinaryResourceImpl;
+import org.eclipse.xtext.common.types.JvmIdentifiableElement;
+import org.eclipse.xtext.resource.persistence.ResourceStorageLoadable;
+import org.eclipse.xtext.resource.persistence.StorageAwareResource;
+import org.eclipse.xtext.xbase.compiler.DocumentationAdapter;
+import org.eclipse.xtext.xbase.jvmmodel.JvmIdentifiableMetaData;
+import org.eclipse.xtext.xbase.jvmmodel.JvmModelAssociator;
+
+import com.google.common.collect.FluentIterable;
+
+public class BatchLinkableResourceStorageLoadable extends ResourceStorageLoadable {
+
+	public BatchLinkableResourceStorageLoadable(InputStream in, boolean storeNodeModel) {
+		super(in, storeNodeModel);
+	}
+
+	@Override
+	protected void loadEntries(StorageAwareResource resource, ZipInputStream zipIn) throws IOException {
+		super.loadEntries(resource, zipIn);
+		if (resource instanceof BatchLinkableResource) {
+			try {
+				readAssociationsAdapter((BatchLinkableResource) resource, zipIn);
+			} catch(ClassNotFoundException e) {
+				throw new IOException(e);
+			}
+		}
+	}
+
+	@Override
+	protected void handleLoadEObject(InternalEObject loaded, BinaryResourceImpl.EObjectInputStream input)
+			throws IOException {
+		super.handleLoadEObject(loaded, input);
+		if (input.readBoolean()) {
+			DocumentationAdapter documentationAdapter = new DocumentationAdapter(input.readString());
+			loaded.eAdapters().add(documentationAdapter);
+		}
+		if (input.readBoolean()) {
+			JvmIdentifiableMetaData metaData = new JvmIdentifiableMetaData();
+			metaData.setSynthetic(input.readBoolean());
+			loaded.eAdapters().add(metaData);
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	protected void readAssociationsAdapter(BatchLinkableResource resource, ZipInputStream stream) throws IOException, ClassNotFoundException {
```

Comment:
api change?

Reply:
You mean the declared ClassNotFoundException?

Reply:
yes

Reply:
Moved the catch clause into this method.